### DISCUSSION TITLE
Ability to disable extensions

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#23941] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Extensions should be allowed to be disabled via some setting, at both the global and local level, so workspaces can disable plugins they don't need running.

**Comment by kucho** [comment]
+1! I would also love the ability to disable some language servers per workspace.

**Reply by turbocrime** [reply]
this feature is available.

here's the zed documentation of this feature https://zed.dev/docs/configuring-languages#choosing-language-servers

and here's an example, from the documentation of biome's zed extension https://biomejs.dev/reference/zed/#project-based-configuration

i think disabling globally and enabling per-project and language is the correct way to do it, but you can of course do the inverse as you ask, and just disable it per-project.

it seems that `language_servers` is both a root config field and available per-language.

**Comment by hempun10** [comment]
+1

**Comment by notpeter** [comment]
While I appreciate the desire to be able to enable/disable extensions so that workspaces only have the extensions they need, I'm unclear of what the design of such an implementation would look like.

- An extension allow-list in project level settings would prevent using optional extensions (themes, slash command extensions, etc)
- An extensions deny-list at the project level would be inherently verbose and non-exhaustive

I'm opened to exploring ideas for how such a system might be designed.

**Reply by kkafar** [reply]
I just found myself in situation where I use different JavaScript language servers / formatters depending on the project I work on & the configuration I have there. I would love to have opportunity to mark e.g. `biome` as disabled and have it not conflicting with `prettier` in another repo.

In other words: in my use cases opportunity to disable **a particular** extension would be desired - "deny-list" approach would be sufficient. I agree that this approach is verbose, but it's not like disabling extensions is such common use case. I feel like it would be a tool to handle certain edge cases (such as mine) & verbosity is not that much of an issue.

**Reply by turbocrime** [reply]
i found myself in this exact situation. for biome specifically, there's a way to configure the extension to automatically enable/disable based on the presence of the `biome.json` config, which is exactly how the command-line biome works:

https://biomejs.dev/reference/zed/#enable-biome-only-when-biomejson-is-present

*\<~/.config/zed>/settings.json*
```json
{
  "lsp": {
    "biome": {
      "settings": {
        "require_config_file": true
      }
    }
  }
}
```

more generally, the `language_servers` zed config field can enable/disable any LSP globally or per workspace https://zed.dev/docs/configuring-languages#choosing-language-servers

**Reply by kkafar** [reply]
That is great! Thank you.

**Reply by Nico-Mayer** [reply]
I tried this and also disabled biome for each file type with '!biome' in the workspace `settings.json`, but biome still connects to every TypeScript and JSON buffer even though I have disabled it and set `"require_config_file": true`.

**Reply by eugenesvk** [reply]
> * An extension allow-list in project level settings would prevent using optional extensions (themes, slash command extensions, etc)

you could have allow-list by extension type? This way you can only allow some language extensions while doing nothing to the themes

**Comment by mantou132** [comment]
[Extensions can now register ts plugins](https://github.com/zed-industries/zed/issues/22410). When the ts plugins registered by two extensions conflict, I need to disable one of the ts plugins. I think disabling the extension in the project configuration is a good solution.

e.g: gem extension conflicts with vue extension

`@vue/typescript-plugin` in vue invalidates some functions of gem's `ts-gem-plugin`

**Reply by mantou132** [reply]
I've tried overriding the ts-plugin configuration and that doesn't work now

<img width="2560" height="1400" alt="image" src="https://github.com/user-attachments/assets/cfdd710a-c596-43ad-9db5-30e1283ef3da" />

**Reply by mantou132** [reply]
Now the only way to get the gem extension working properly is to uninstall the vue extension
